rust/bin/agent/src/handler/upsert.rs

```diff
@@ -18,147 +18,179 @@ use log::{info, warn};
 use prost::Message;
 use proto::{
     payload::v1::{insert, object, update, upsert},
-    vald::v1::{insert_server::Insert, update_server::Update, upsert_server},
+    vald::v1::upsert_server,
 };
-use std::collections::HashMap;
+use tokio::sync::RwLock;
+use std::{collections::HashMap, sync::Arc};
 use tonic::{Code, Status};
 use tonic_types::{ErrorDetails, FieldViolation, StatusExt};
 
+use super::update::update as update_fn;
+use super::insert::insert as insert_fn;
+use super::common::bidirectional_stream;
+
+async fn upsert(
+    s: Arc<RwLock<dyn algorithm::ANN>>,
+    resource_type: &str,
+    api_name: &str,
+    name: &str,
+    ip: &str,
+    request: &upsert::Request
+) -> Result<object::Location, Status> {
+    let config = match request.config.clone() {
+        Some(cfg) => cfg,
+        None => return Err(Status::invalid_argument("Missing configuration in request")),
+    };
+    let hostname = cargo::util::hostname()?;
+    let domain = hostname.to_str().unwrap();
+    {
+        let vec = match request.vector.clone() {
+            Some(v) => v,
+            None => return Err(Status::invalid_argument("Missing vector in request")),
+        };
+        let s_inner = s.read().await;
+        let uuid = vec.id.clone();
```

Comment:
do we need clone id?
seems vec is already cloned, ownership of id is already in scope.

Reply:
Maybe, this seems to be because “partially move value” is occurred in combination with the following comment.
These combination codes enable to compile:
```
let uuid = vec.id;
...
vector: request.vector.clone(),
```
or
```
let uuid = vec.id.clone();
...
vector: Some(vec),
```